### PR TITLE
Improve error messaging

### LINKED ISSUE #151: Improve bridge error handling
#### Description

- Add input validation on the Amount input and show errors under the amount input rather then in the error box.

### PR DESCRIPTION
**Pull request by alexkeating**
#### Description

- Convert input to react-hook-form validation
- Check if negative number
- Check if input is a number
- Make error box overflow on auto

Closes #151

### CODE REVIEW

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* main
  * **PR #154** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/154?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #157** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/157?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #156** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/156?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/157?utm_source=stack-comment).

File: config.ts [hunk 1]
```diff
@@ -2,6 +2,10 @@ import { goerli, optimismGoerli } from 'wagmi/chains';
 
 export const chains = [goerli, optimismGoerli];
 
+declare global {
```
Comment (wildmolasses) [inline config.ts:5]: hmm, this should probably be relocated to a types file. in fact, can we remove and just make the coercion to number explicit? i'll give an example later in review

File: pages/[id]/bridge.tsx [hunk 2]
```diff
@@ -44,9 +49,19 @@ const Bridge = () => {
...
+    formState: { errors, isValid },
+    watch,
+  } = useForm<FormData>({ mode: 'onChange' });
+
   // State for amount input
-  const [amount, setAmount] = useState<string>('0');
-  const rawAmount = parseUnits(amount, l1.token?.decimals || 18);
+  const amount = watch('amount', '0');
+  const rawAmount = parseUnits(
+    Boolean(errors.amount?.message) || isNaN(amount) ? '0' : amount,
```
Comment (wildmolasses) [inline pages/[id]/bridge.tsx:62]: explicit coercion to number

```suggestion
    Boolean(errors.amount?.message) || isNaN(+amount) ? '0' : amount,
```

File: pages/[id]/bridge.tsx [hunk 3]
```diff
@@ -261,37 +275,48 @@ const Bridge = () => {
...
                   placeholder="0.00"
                   defaultValue=""
-                  aria-invalid={errorType === ErrorType.ERC20AmountError}
+                  aria-invalid={errorType === ErrorType.ERC20AmountError || !isValid}
                   aria-describedby="amount"
-                  onChange={handleInputChange}
+                  {...register('amount', {
+                    validate: async (value) => {
+                      // isNan coerces to a number
+                      if (isNaN(value)) {
```
Comment (wildmolasses) [inline pages/[id]/bridge.tsx:297]:
```suggestion
                      if (isNaN(+value)) {
```

File: pages/[id]/bridge.tsx [hunk 4]
```diff
@@ -261,37 +275,48 @@ const Bridge = () => {
...
                   )}
                   placeholder="0.00"
                   defaultValue=""
-                  aria-invalid={errorType === ErrorType.ERC20AmountError}
+                  aria-invalid={errorType === ErrorType.ERC20AmountError || !isValid}
                   aria-describedby="amount"
-                  onChange={handleInputChange}
+                  {...register('amount', {
+                    validate: async (value) => {
+                      // isNan coerces to a number
```
Comment (wildmolasses) [inline pages/[id]/bridge.tsx:296]:
```suggestion
```

File: pages/[id]/bridge.tsx [hunk 5]
```diff
@@ -130,7 +144,7 @@ const Bridge = () => {
     isLoading: bridgeToL1IsLoading,
     error: bridgeToL1Error,
   } = useEasyWrite({
-    enabled: isNonZeroInput && bridgeTarget === BridgeTarget.L1 && source.chain.id === chain?.id,
```
Comment (wildmolasses) [inline pages/[id]/bridge.tsx:133]: nit: why remove here? do we want this hook to run when the input is 0?

File: pages/[id]/bridge.tsx [hunk 6]
```diff
@@ -108,7 +123,6 @@ const Bridge = () => {
     error: bridgeToL2Error,
   } = useEasyWrite({
     enabled:
-      isNonZeroInput &&
```
Comment (wildmolasses) [inline pages/[id]/bridge.tsx:111]: nit: why remove here? do we want this hook to run when the input is 0?

**Review by wildmolasses — APPROVED** [review]
works for me; a few nits to consider, otherwise good to merge

File: config.ts (hunk 1 above)
Comment (alexkeating) [reply config.ts:5]: yes, an example would be much appreciated. I tried with `as`, but it didn't work.

File: pages/[id]/bridge.tsx (hunk 5 above)
Comment (alexkeating) [reply pages/[id]/bridge.tsx:133]: Good catch,  there was a latency when testing, but now everything seems to be good.

File: config.ts (hunk 1 above)
Comment (alexkeating) [reply config.ts]: https://github.com/ScopeLift/l2-flexible-voting-frontend/pull/157/commits/283e41df1bf91e1a6eba8e6be69699ab7fa38a85

File: pages/[id]/bridge.tsx (hunk 2 above)
Comment (alexkeating) [reply pages/[id]/bridge.tsx]: https://github.com/ScopeLift/l2-flexible-voting-frontend/pull/157/commits/283e41df1bf91e1a6eba8e6be69699ab7fa38a85

File: pages/[id]/bridge.tsx (hunk 3 above)
Comment (alexkeating) [reply pages/[id]/bridge.tsx]: https://github.com/ScopeLift/l2-flexible-voting-frontend/pull/157/commits/283e41df1bf91e1a6eba8e6be69699ab7fa38a85

File: pages/[id]/bridge.tsx (hunk 4 above)
Comment (alexkeating) [reply pages/[id]/bridge.tsx]: https://github.com/ScopeLift/l2-flexible-voting-frontend/pull/157/commits/283e41df1bf91e1a6eba8e6be69699ab7fa38a85

File: pages/[id]/bridge.tsx (hunk 5 above)
Comment (alexkeating) [reply pages/[id]/bridge.tsx:133]: https://github.com/ScopeLift/l2-flexible-voting-frontend/pull/157/commits/283e41df1bf91e1a6eba8e6be69699ab7fa38a85

File: pages/[id]/bridge.tsx (hunk 6 above)
Comment (alexkeating) [reply pages/[id]/bridge.tsx:111]: https://github.com/ScopeLift/l2-flexible-voting-frontend/pull/157/commits/283e41df1bf91e1a6eba8e6be69699ab7fa38a85

**Pull request merged**